### PR TITLE
make it possible to pass an instance of an integration class

### PR DESCRIPTION
**Pull request by lxbdr**
I created a reusable integration class for Elementor repeater fields where I need to add specifics by passing constructor params which is currently not possible. This addition makes it possible to pass an instance instead of a class string.

### CODE REVIEW

File: src/class-wpml-elementor-translatable-nodes.php [hunk 1]
```diff
@@ -74,7 +74,11 @@ public function get( $node_id, $element ) {
 				if ( isset( $node_data['integration-class'] ) ) {
 					foreach ( $this->get_integration_classes( $node_data ) as $class ) {
 						try {
-							$node    = new $class();
+							if($class instanceof \WPML_Elementor_Module_With_Items){
```
Comment (strategio) [inline src/class-wpml-elementor-translatable-nodes.php]: @lxbdr, could you please add spaces around the parenthesis: `if ( $class instanceof \WPML_Elementor_Module_With_Items ) {`.

We are trying to follow the WP coding standards (https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#brace-style).

File: src/class-wpml-elementor-translatable-nodes.php [hunk 2]
```diff
@@ -116,7 +120,11 @@ public function update( $node_id, $element, WPML_PB_String $string ) {
 				if ( isset( $node_data['integration-class'] ) ) {
 					foreach ( $this->get_integration_classes( $node_data ) as $class ) {
 						try {
-							$node = new $class();
+							if($class instanceof \WPML_Elementor_Module_With_Items){
```
Comment (strategio) [inline src/class-wpml-elementor-translatable-nodes.php]: Same here please @lxbdr.

**Review by strategio — CHANGES_REQUESTED** [review]
Thanks for your contribution @lxbdr :)

I have a couple of formatting comments, but your change is legit in my opinion. Would you be able to cover this changes with a test? If not, no worries we will do it.

**Comment by lxbdr** [comment]
@strategio thanks for the review. I added spaces to brackets. I tested the proposed changes on my setup and can confirm that it works.

Unfortunately I don't have any experience with unit tests so far but I looked at the existing code and tried to put something together. As I altered the method for getting and updating nodes I copied the tests for getting and updating single repeater fields so I think this should cover the changes - hope this is what you were looking for :)

**Comment by strategio** [comment]
Thanks @lxbdr, and congratulations because your tests are passing :)

However, I see only partial formatting adjustments (and it's not "symetric"):
- `if( $class instanceof \WPML_Elementor_Module_With_Items ){`
- `if ( $class instanceof \WPML_Elementor_Module_With_Items ){`

Could you replace it by `if ( $class instanceof \WPML_Elementor_Module_With_Items ) {` ?

Meanwhile I will ask for another reviewer (@OnTheGoSystems/wpml).

**Review by strategio — APPROVED** [review]
It sounds good @lxbdr, thanks! I will ask for a second review and then I will ask you to squash your commits.

@OnTheGoSystems/wpml one more review please!

File: src/class-wpml-elementor-translatable-nodes.php [hunk 3]
```diff
@@ -74,7 +74,11 @@ public function get( $node_id, $element ) {
 				if ( isset( $node_data['integration-class'] ) ) {
 					foreach ( $this->get_integration_classes( $node_data ) as $class ) {
 						try {
-							$node    = new $class();
+							if ( $class instanceof \WPML_Elementor_Module_With_Items ) {
+								$node = $class;
+							} else {
+								$node = new $class();
+							}
```
Comment (andreasciamanna) [inline src/class-wpml-elementor-translatable-nodes.php:81]: @strategio I'm not a huge fan of this sort of changes in existing API.

I do understand the reasons, but every time we do that, we make our API less deterministic.

Anyway, this is a month old PR: is it still valid?

Comment (lxbdr) [reply src/class-wpml-elementor-translatable-nodes.php:81]: Yes this PR is still valid.

I understand your argument and with checking whether it is an instance of the expected class it stays deterministic. I tried hard to find a different approach, with anonymous classes, using static properties, etc. but haven't found a way to make it work other than this...

Comment (strategio) [reply src/class-wpml-elementor-translatable-nodes.php:81]: Personnally, I think this solution is still valid. As an alternative, we could also allow to pass a factory, but it requires more changes I believe (and you won't be a huge fan of this either).

I'll leave you the final word.

**Review by andreasciamanna — APPROVED** [review]

**Comment by andreasciamanna** [comment]
This PR still requires a refresh. Probably, just a rebase.

**Comment by lxbdr** [comment]
@strategio just wanted to follow up on this one... Could you please do the 'squashing' or what is necessary so the PR can be merged?

**Comment by strategio** [comment]
Hi @lxbdr! Thank you for reminding this PR, it was under the radar.

As I can't push anything on your repository, I created a squashed/rebased branch on ours, and opened a new PR (see https://github.com/OnTheGoSystems/wpml-page-builders-elementor/pull/89).

**Comment by dimitrismitsis** [comment]
This has been merged and it will be included in upcoming WPML bundle releases.

**Pull request closed without merging**